### DISCUSSION TITLE
How to get gateway messages/errors on front-end

### DISCUSSION THREAD

**Discussion by jrrdnx** [digital-pros/commerce-authorize#26] [Q&A] [answered]
I have this plugin implemented and mostly working, but am trying to figure out how to display error messages from the gateway to the user on the front-end.

In this example, I've exceeded the maximum amount allowed (I know, extreme edge case, but just using as an example) but the array returned by `paymentForm.getErrors()` is empty. I can see the transaction details on the attempted payment in the admin but that's it.

Admin transaction details:
![Screen Shot 2022-01-11 at 08 08 00](https://user-images.githubusercontent.com/4584671/148948599-05cbd24f-b472-4183-adb5-4f0fefec8570.png)

Front-end code:
```
<ul class="errors">
    {%- if paymentForm is defined -%}
        <pre>{{ dump(paymentForm) }}</pre>
        {% for key, errors in paymentForm.getErrors() %}
            {% for error in errors %}
                <li>
                    <strong>{{ key }}</strong> {{ error }}
                </li>
            {% endfor %}
        {% endfor %}
    {%- endif -%}
</ul>
```
`paymentForm` object on front-end:
```
object(digitalpros\commerce\authorize\models\AuthorizePaymentForm)#1484 (16) {
  ["customerProfile"]=>
  NULL
  ["firstName"]=>
  string(7) "testFirst"
  ["lastName"]=>
  string(6) "testLast"
  ["number"]=>
  string(16) "4111111111111111"
  ["month"]=>
  string(1) "1"
  ["year"]=>
  string(4) "2027"
  ["cvv"]=>
  string(3) "456"
  ["token"]=>
  NULL
  ["expiry"]=>
  NULL
  ["threeDSecure"]=>
  bool(false)
  ["_errors":"yii\base\Model":private]=>
  array(0) {
  }
  ["_validators":"yii\base\Model":private]=>
  object(ArrayObject)#1959 (1) {
    ...
  }
  ["_scenario":"yii\base\Model":private]=>
  string(7) "default"
  ["_events":"yii\base\Component":private]=>
  array(0) {
  }
  ["_eventWildcards":"yii\base\Component":private]=>
  array(0) {
  }
  ["_behaviors":"yii\base\Component":private]=>
  array(0) {
  }
}
```

So it appears that there is an error message from Authorize.net in `transactionResponse.errors`, but I'm not sure how I can access that on the front-end. Any help is greatly appreciated, thank you!

**Comment by benlobach** [comment, marked as answer]
@jrrdnx, great question! We don't override anything within the omnipay payment method, which is why there isn't an error in the paymentForm directly. You should see a generic error in the flashErrors, which explains that the payment wasn't successful.

`{% set flashError = craft.app.session.getFlash('error') %}`

If you'd like to output the transaction details specifically, you could watch for this error within your payment template, and then loop through the transactions on the cart, and pull transaction message and details from the transaction itself. 

That would look something like the following:

```
{% if flashError %}
      {% for transaction in cart.getLastTransaction() %}
          {{ transaction.message }}
      {% endfor %}
{% endif %}
```
Here's the transaction model, in case that helps!

https://docs.craftcms.com/commerce/api/v3/craft-commerce-models-transaction.html#public-properties 

_(edited for clarity and formatting, and to remove the invalid function)_

**Reply by jrrdnx** [reply]
Thanks @benlobach !

`hasTransactions()` is throwing an invalid method error for me, but my `cart` object does have a `lastTransaction` property that I'm able to access so I've consolidated my check to:

```
{%- if flashError and cart.lastTransaction -%}
    {{ cart.lastTransaction.message }}
{%- endif -%}
```

**Reply by benlobach** [reply]
Thanks for letting me know! I may have been looking at older documentation when I saw that function. Glad you were able to work it out with the the flashError!
